Benchmark differenciation

MarineLasbleis opened the pull request:
I still need to write the manual part! I'll update soon.

MarineLasbleis commented:
I update the files and added a section in the manual. 
I did not managed to see the output of the manual, as I got "I couldn't open style file abbrvurl.bst" (which I assume only comes from my pdflatex installation). I'll have a look, but at least the manual addition looks good when I was compiling it on overleaf, so equations, figures, etc. should be OK. 

So, I think this is ready for someone to have a look!

gassmoeller commented on benchmarks/differenciation/python/aspect_python.py:
```diff
@@ -0,0 +1,134 @@
+import numpy as np
```
Is the content of this files also in #3088? Is there a way to use only one (when both are merged)?

MarineLasbleis replied:
There would be a way, by packaging the python code into a pip package. However, this seems to be a little bit extreme for a couple of functions. So far, it is the same file than in the PR #3088, and using links to it is a little bit weird. I could remove it here and ask people to copy paste the file here...
for now, I think it's simpler to use a file with the functions here (especially as there is no tests on the python code, so even with a python package, there would be no way to be sure that nothing has been broken later and that these functions no longer work)

gassmoeller commented on doc/manual/manual.tex:
```diff
@@ -10818,6 +10818,125 @@ \subsubsection{Yamauchi \& Takei anelastic shear wave velocity-temperature conve
+
+\textit{This section was contributed by Marine Lasbleis.}
+
+Here we use \aspect{} to mode the compaction of a simple column with a mixture of liquid and solid using two-phase flow equations. 
```
```suggestion
Here we use \aspect{} to model the compaction of a simple column with a mixture of liquid and solid using two-phase flow equations. 
```

gassmoeller commented on doc/manual/manual.tex:
```diff
@@ -10818,6 +10818,125 @@ \subsubsection{Yamauchi \& Takei anelastic shear wave velocity-temperature conve
+and the liquid viscosity negligible compared to the solid one. 
+Both phases are considered to be viscously deformed by the compaction, and the liquid is lighter than the solid. 
+
+The plugin and parameter files can be found at \url{benchmarks/differenciation/melt_srb.cc} and \url{benchmarks/differenciation/diff_adim.prm}. 
```
need to adjust the directory names here and in the line below

MarineLasbleis replied:
done

gassmoeller commented on doc/manual/manual.tex:
```diff
@@ -10818,6 +10818,125 @@ \subsubsection{Yamauchi \& Takei anelastic shear wave velocity-temperature conve
+\end{equation}
+and is constant with the porosity. 
+
+The global melt model is set up with compressibility, which we will need to remove by setting the parameters to 0. 
```
```suggestion
The global melt model is originally set up with compressibility, which we will need to remove by setting the corresponding parameters to 0. 
```

gassmoeller commented on doc/manual/manual.tex:
```diff
@@ -10818,6 +10818,125 @@ \subsubsection{Yamauchi \& Takei anelastic shear wave velocity-temperature conve
+
+The global melt model is set up with compressibility, which we will need to remove by setting the parameters to 0. 
+We are also neglecting the effects of temperature on the parameters and setting the temperature field to 0 everywhere. 
+The viscosities tends to 0 or infinity around $\phi=0$ and $\phi=1$, which is not possible to implement in numerical model. 
```
```suggestion
The viscosities tends to 0 or infinity around $\phi=0$ and $\phi=1$, which is not possible to implement in numerical models. 
```

gassmoeller commented on doc/manual/manual.tex:
```diff
@@ -10818,6 +10818,125 @@ \subsubsection{Yamauchi \& Takei anelastic shear wave velocity-temperature conve
+The global melt model is set up with compressibility, which we will need to remove by setting the parameters to 0. 
+We are also neglecting the effects of temperature on the parameters and setting the temperature field to 0 everywhere. 
+The viscosities tends to 0 or infinity around $\phi=0$ and $\phi=1$, which is not possible to implement in numerical model. 
+We set a maximum viscosity variation due to porosity of $10^6$.
```
```suggestion
We therefore set a maximum viscosity variation due to porosity of $10^6$.
```

gassmoeller commented on doc/manual/manual.tex:
```diff
@@ -10818,6 +10818,125 @@ \subsubsection{Yamauchi \& Takei anelastic shear wave velocity-temperature conve
+We set a maximum viscosity variation due to porosity of $10^6$.
+
+We defined a new material model which incorporate all of these modifications, based on the material model global melt. 
+The plugin \url{benchmarks/differenciation/melt_srb.cc} is used as an additional library and needs to be compiled befre running the parameter file. 
```
```suggestion
The plugin \url{benchmarks/differenciation/melt_srb.cc} is used as an additional library and needs to be compiled before running the parameter file. 
```

gassmoeller commented on doc/manual/manual.tex:
```diff
@@ -10818,6 +10818,125 @@ \subsubsection{Yamauchi \& Takei anelastic shear wave velocity-temperature conve
+We defined a new material model which incorporate all of these modifications, based on the material model global melt. 
+The plugin \url{benchmarks/differenciation/melt_srb.cc} is used as an additional library and needs to be compiled befre running the parameter file. 
+
+\paragraph{Set-up}
```
```suggestion
\paragraph{Model Setup}
```

gassmoeller commented on doc/manual/manual.tex:
```diff
@@ -10818,6 +10818,125 @@ \subsubsection{Yamauchi \& Takei anelastic shear wave velocity-temperature conve
+\paragraph{Set-up}
+
+We set a column of width one cell, with 200 cells in the y-direction. 
+The total size of the system is 1 and all parameters are put to 0 or 1 to have a dimensionless setting. 
```
```suggestion
The total area of the system is 1 and all parameters are put to 0 or 1 to have a dimensionless setting. 
```

MarineLasbleis replied:
it's the total height. Modified.

gassmoeller commented on doc/manual/manual.tex:
```diff
@@ -10818,6 +10818,125 @@ \subsubsection{Yamauchi \& Takei anelastic shear wave velocity-temperature conve
+We use a compaction length of 1 and 0.2 for this benchmark (ones need to set the fluid viscosity to 25 to change this value from 1 to 0.2), 
+and we start with a constant porosity profile of 30\%.
+
+The parameter file is available at \url{benchmarks/differenciation/diff_adim.prm}. 
```
change directory name here

gassmoeller commented on doc/manual/manual.tex:
```diff
@@ -10818,6 +10818,125 @@ \subsubsection{Yamauchi \& Takei anelastic shear wave velocity-temperature conve
+\end{equation}
+
+Comparison between the analytical solution and our runs for two compaction lengths (1 and 0.2) are shown on figure \ref{fig:analytic}. 
+The agreement is good, with a slight discrepancy at the top, likely due to the top boundary condition not exactly defines at $y=1$.   
```
```suggestion
The agreement is good, with a slight discrepancy at the top, likely due to the top boundary condition not exactly defined at $y=1$.   
```

gassmoeller commented on doc/manual/manual.tex:
```diff
@@ -10818,6 +10818,125 @@ \subsubsection{Yamauchi \& Takei anelastic shear wave velocity-temperature conve
+
+\paragraph{Evolution}
+
+The porosity profiles evolves such that the liquid is expelled to the top and the matrix compacts at the bottom. 
```
```suggestion
The porosity profile evolves such that the liquid is expelled to the top and the matrix compacts at the bottom. 
```

gassmoeller commented on doc/manual/manual.tex:
```diff
@@ -10818,6 +10818,125 @@ \subsubsection{Yamauchi \& Takei anelastic shear wave velocity-temperature conve
+
+
+The porosity and matrix velocity evolution are shown on figure \ref{fig:evolution}. 
+Results are similar to \cite{SRPHD07}, except for porosity close to 1, where \aspect{} do not converge correctly and overestimate the segregation. 
```
```suggestion
Results are similar to \cite{SRPHD07}, except for porosity close to 1, where \aspect{} does not converge correctly and overestimates the segregation. 
```

gassmoeller commented on doc/manual/manual.tex:
```diff
@@ -10818,6 +10818,125 @@ \subsubsection{Yamauchi \& Takei anelastic shear wave velocity-temperature conve
+
+The porosity and matrix velocity evolution are shown on figure \ref{fig:evolution}. 
+Results are similar to \cite{SRPHD07}, except for porosity close to 1, where \aspect{} do not converge correctly and overestimate the segregation. 
+Since the model was designed to solve for very low porosity, and that the equations should not be physically valid for porosity high enough than the matrix is no longer connected (the highest threshold usually considered is about 40\%, corresponding to a system made of spheres with a random close packing)
```
This sentence is not complete. You likely want to say the above deviation is expected, or at least explainable?

MarineLasbleis replied:
modified

gassmoeller commented on benchmarks/differenciation/melt_srb.cc:
```diff
@@ -0,0 +1,698 @@
+    using namespace dealii;
+
+    /**
+     * A material model that implements a simple formulation of the
```
This is likely copy-pasted from the melt global model, please replace by a short sentence saying this is a modification of melt global for the benchmark you are running. Also leave a sentence what you changed from melt global.

gassmoeller commented on benchmarks/differenciation/melt_srb.cc:
```diff
@@ -0,0 +1,698 @@
+              double porosity = std::max(in.composition[i][porosity_idx],0.0);
+
+              melt_out->fluid_viscosities[i] = eta_f;
+              melt_out->permeabilities[i] = reference_permeability * std::pow(porosity,2); // * std::pow(1.0-porosity,2);
```
Please remove the commented code, unless it is important. Leaving commented code around is bad style and causes confusion in the future (plus we do not need it, because we have git for the history).

MarineLasbleis replied:
done (and I added info because this is one of the places where I modified the initial melt model)

MarineLasbleis commented:
About the difference between the global melt and the plugin in the benchmark: I needed to change the functions for the viscosities and for the permeability, which are both in the middle of the material model. Ideally, they would be functions in the material model that can be overwritten in derived classes. But I am not sure exactly how derived classes work in C++ (sorry, I started C++ for the hackathon), and when I tried to do a derived class from the global melt (to use at least some of the already-written functions), it did not work. Apparently, I can't just use the same way it was done for the inner core plugin (based on the material model simpler), and I am not sure why. 
The modifications of the viscosities and permeability functions seem to be interested for almost all people working on two-phase flow, so I may try to re-make a material model that is more adaptable later (ideally, I guess, with the functions used in GlobalMelt being the defaults, to ensure that nothing is broken)

And I'll work on the typos and changes in names later this week!

MarineLasbleis commented on benchmarks/differenciation/CMakeLists.txt:
```diff
@@ -0,0 +1,40 @@
+# Copyright (C) 2015 - 2019 by the authors of the ASPECT code.
```
done

MarineLasbleis commented on benchmarks/differenciation/diff_adim.prm:
```diff
@@ -0,0 +1,246 @@
+# Benchmark for the equations of SRB 
```
done

MarineLasbleis commented on benchmarks/differenciation/diff_adim.prm:
```diff
@@ -0,0 +1,246 @@
+# Benchmark for the equations of SRB 
+# See Sramek PhD thesis 2007 p48
```
done

MarineLasbleis commented on benchmarks/differenciation/diff_adim.prm:
```diff
@@ -0,0 +1,246 @@
+
+# We use here a dimensionless formulation.
+
+set Additional shared libraries            = ./libmelt_srb.so 
```
done

MarineLasbleis commented on benchmarks/differenciation/diff_adim.prm:
```diff
@@ -0,0 +1,246 @@
+set Additional shared libraries            = ./libmelt_srb.so 
+
+set Dimension                              = 2
+set Output directory                       = output-srb
```
done

MarineLasbleis commented on benchmarks/differenciation/CMakeLists.txt:
```diff
@@ -0,0 +1,40 @@
+SET(TARGET "melt_srb")
+PROJECT(${TARGET})
+
+ADD_LIBRARY(${TARGET} SHARED melt_srb.cc)
```
it's the author names. I changed it

MarineLasbleis commented:
I think I went through all the remarks. Let me know if I forget one of them!

gassmoeller commented on benchmarks/melt_column_differenciation/melt_SramekRicardBercovici2007.cc:
```diff
@@ -0,0 +1,703 @@
+    /**
+     * This material model is c/c from the melt global model
+     * to be used in the benchmarck melt_column_differenciation. 
+     * From the initial model, the permeability, the bulk and the shear viscosity
```
```suggestion
     * From the original material model, the permeability, the bulk and the shear viscosity
```

gassmoeller commented on benchmarks/melt_column_differenciation/melt_SramekRicardBercovici2007.cc:
```diff
@@ -0,0 +1,703 @@
+     * to be used in the benchmarck melt_column_differenciation. 
+     * From the initial model, the permeability, the bulk and the shear viscosity
+     * have been modified to solve the same set of equation than in 
+     * Sramek, Ricard and Bercovici 2007 (without fusion/cristallization)
```
```suggestion
     * Sramek, Ricard and Bercovici 2007 (without fusion/crystallization)
```

gassmoeller commented on benchmarks/melt_column_differenciation/melt_SramekRicardBercovici2007.cc:
```diff
@@ -0,0 +1,703 @@
+  You should have received a copy of the GNU General Public License
+  along with ASPECT; see the file LICENSE.  If not see
+  <http://www.gnu.org/licenses/>.
+*/
```
Please remove this copyright notice. Each file in ASPECT requires exactly one of these notices at the very top of the file.

gassmoeller commented on benchmarks/melt_column_differenciation/melt_SramekRicardBercovici2007.cc:
```diff
@@ -0,0 +1,703 @@
+  {
+    ASPECT_REGISTER_MATERIAL_MODEL(MeltSRB,
+                                   "melt srb",
+                                   "A material model  modified from melt global"
```
```suggestion
                                   "A material model modified from melt global"
```

gassmoeller commented on doc/manual/manual.tex:
```diff
@@ -10818,6 +10818,126 @@ \subsubsection{Yamauchi \& Takei anelastic shear wave velocity-temperature conve
+and the liquid viscosity negligible compared to the solid one. 
+Both phases are considered to be viscously deformed by the compaction, and the liquid is lighter than the solid. 
+
+The plugin and parameter files can be found at \url{benchmarks/melt_column_differenciation/melt_SramekRicardBercovici2007.cc} and \url{benchmarks/melt_column_differenciation/diff_adim.prm}. 
```
need to adjust the directory name

gassmoeller commented on doc/manual/manual.tex:
```diff
@@ -10818,6 +10818,126 @@ \subsubsection{Yamauchi \& Takei anelastic shear wave velocity-temperature conve
+Both phases are considered to be viscously deformed by the compaction, and the liquid is lighter than the solid. 
+
+The plugin and parameter files can be found at \url{benchmarks/melt_column_differenciation/melt_SramekRicardBercovici2007.cc} and \url{benchmarks/melt_column_differenciation/diff_adim.prm}. 
+A jupyter notebook (Python 3) for recreating the different figures shown here is provided in \url{benchmarks/melt_column_differenciation/python}. 
```
and here

gassmoeller commented on doc/manual/manual.tex:
```diff
@@ -10818,6 +10818,126 @@ \subsubsection{Yamauchi \& Takei anelastic shear wave velocity-temperature conve
+We therefore set a maximum viscosity variation due to porosity of $10^6$.
+
+We defined a new material model which incorporate all of these modifications, based on the material model global melt. 
+The plugin \url{benchmarks/melt_column_differenciation/melt_srb.cc} is used as an additional library and needs to be compiled before running the parameter file. 
```
and here

gassmoeller commented on doc/manual/manual.tex:
```diff
@@ -10818,6 +10818,126 @@ \subsubsection{Yamauchi \& Takei anelastic shear wave velocity-temperature conve
+We use a compaction length of 1 and 0.2 for this benchmark (ones need to set the fluid viscosity to 25 to change this value from 1 to 0.2), 
+and we start with a constant porosity profile of 30\%.
+
+The parameter file is available at \url{benchmarks/melt_column_differenciation/diff_adim.prm}. 
```
and here

gassmoeller commented on doc/manual/manual.tex:
```diff
@@ -10818,6 +10818,126 @@ \subsubsection{Yamauchi \& Takei anelastic shear wave velocity-temperature conve
+
+The porosity and matrix velocity evolution are shown on figure \ref{fig:evolution}. 
+Results are similar to \cite{SRPHD07}, except for porosity close to 1, where \aspect{} does not converge correctly and overestimates the segregation. 
+The model was designed to solve for very low porosity, and the equations are not be physically valid for porosity high enough -- the matrix is then no longer connected -- then this result was expected.  
```
```suggestion
The model was designed to solve for very low porosity, and the equations are not be physically valid for porosity high enough -- the matrix is then no longer connected -- therefore this result was expected.  
```

MarineLasbleis commented:
Sorry for the typo in the folder name! I corrected all of the remarks, thank you very much. 
I realized there is no test associated, should I set one? 
For the file in doc/modules/changes, what is the date I should use? I added one, but used today's date. Should I changed it to the date I did the pull request initially? I am guessing it does not matter so much, except if a different date may appear somewhere.

gassmoeller commented:
>  I realized there is no test associated, should I set one?

Yes please, I keep forgetting to ask for one.

> For the file in doc/modules/changes, what is the date I should use? 

That does not matter very much. The date you started working on this, the date you opened the pull request, or today, any is fine. The date is mostly there to create unique file names, and to order the files approximately correctly.

The changes look good, but you will need to run `make indent` for indentation.

MarineLasbleis commented:
I changed the indent, and added tests files (i was waiting for the test to run here to copy the results)
I don't really understand why the "commit cannot be built". The error seems to come from docker?

gassmoeller commented:
The CIG tester has some problems at the moment, the deciding error message from the tjhei tester is this:
```
CMake Error at CMakeLists.txt:280 (MESSAGE):

  test melt_column_differenciation.prm is missing a folder with reference data

```

tjhei commented:
Many of your commits are not created with your correct email address configured in git (the commits with the grey icon). Do you know how to rewrite your history to fix this? You will need to first set your git email address.

tjhei commented on tests/melt_column_differenciation.prm:
```diff
@@ -0,0 +1,17 @@
+set Dimension = 2
+
+include $ASPECT_SOURCE_DIR/benchmarks//melt_column_differentiation/diff_adim.prm
```
```suggestion
include $ASPECT_SOURCE_DIR/benchmarks/melt_column_differentiation/diff_adim.prm
```

tjhei commented on doc/modules/changes/20190705_lasbleis:
```diff
@@ -0,0 +1,5 @@
+New: Additional benchmark to test the 2-phase flow equations 
```
```suggestion
New: Additional benchmark "melt_column_differentiation" to test the 2-phase flow equations 
```

tjhei commented on tests/melt_column_differenciation.prm:
```diff
@@ -0,0 +1,17 @@
+set Dimension = 2
```
can you please rename the files to be differentiation not differenciation?

MarineLasbleis commented:
Sorry @tjhei , I have no idea why some of the commits do not have the proper name (and no idea how to change history for that...). It may be that my work computer do not have proper github settings as I'm using mostly with gitlab. Any tutorial on how to modify history for that?

gassmoeller commented:
Hi Marine, sorry for coming back so late to this PR. I took the liberty of addressing the remaining comments and fixing up the history (using a `git rebase -i master`). To upload them I had to force push to this branch, so currently pulling or pushing to this branch will not work for you. Hopefully this is more or less ready as is (asking for a review :-)) so this will not be a problem.

gassmoeller commented:
@MarineLasbleis it looks like there are some figures missing for the manual entry. Do you still have them? Otherwise we will need to recreate them.

MarineLasbleis commented:
Hi Rene, and sorry, I also forget this was pending... 
I found the figures! However, since you said I may not be able to push or pull this... Should I send them to you by emails? 
I also realized looking at the manual that I put them as 
\includegraphics[width=0.49\textwidth]{cookbooks/benchmarks/delta_1_analytical.pdf}
Which means they would have to be in the benchmarks directory and not in a subfolder of the benchmark directory... That may not be wise? Where should the figures be for the manual? I think it means we need to change the entries in the manual to 
\includegraphics[width=0.49\textwidth]{cookbooks/benchmarks/melt_column_differentiation/delta_1_analytical.pdf}

Any chance I can do that myself? Let me know what is best. So far, everything is on my computer at work (where I am not, I am out of work until June, but I have remote access and can do that, but maybe not "very" fast!)

gassmoeller commented:
Hi Marine,
Just adding the figures and fixing the paths should be quick, so sending me an email is probably easiest, after that we do not need this branch anymore. If you want to fix your local branch instead you can do the following:
```
git checkout master
git branch -D benchmark_diff
git remote update
git checkout origin/benchmark_diff
git checkout -b benchmark_diff
```
This deletes your local copy of this branch, then gets this online version of the branch.

> and can do that, but maybe not "very" fast!)

no worries, you can send me the files whenever you have the time.

tjhei commented:
@MarineLasbleis do you have plans to finish this PR?

MarineLasbleis commented:
Sorry, I completely forgot about it. 
I no longer have the files (or the means to use ASPECT actually). 
So in short : no, not really.

gassmoeller commented:
Then let's close this PR. Maybe we come back to it later in case someone is interested in picking it up again.
